cli/internal/runsummary/spaces.go

```diff
@@ -1,13 +1,235 @@
+	run      *spaceRun
+	wg       sync.WaitGroup
+	spaceID  string
+	enabled  bool
+}
+
+type spaceRun struct {
+	ID      string
+	URL     string
+	created chan struct{} // a signal that the run has completed
+}
+
+func newSpacesClient(spaceID string, api *client.APIClient, ui cli.Ui, rsm *Meta) *spacesClient {
+	return &spacesClient{
+		api:      api,
+		ui:       ui,
+		rsm:      rsm,
+		spaceID:  spaceID,
+		enabled:  spaceID != "",
+		requests: make(chan *spaceRequest), // TODO: give this a size based on tasks
+		// Set a default, empty one here, so we'll have something downstream and not a segfault
+		run: &spaceRun{
+			created: make(chan struct{}, 1),
+		},
+	}
+}
+
+// Start receiving and processing requests in 8 goroutines
+// There is an additional marker (protected by a mutex) that indicates
+// when the first request is done. All other requests are blocked on that one.
+// This first request is the POST /run request. We need to block on it because
+// the response contains the run ID from the server, which we need to construct the
+// URLs of subsequent requests.
+func (c *spacesClient) start() {
+	if !c.enabled {
+		return
+	}
+
+	mu := sync.Mutex{}
+	firstReqDone := false
+	processors := 8
+	for i := 0; i < processors; i++ {
+		c.wg.Add(1)
+		go func() {
+			defer c.wg.Done()
+			for req := range c.requests {
+				// since we have multiple processors, we want to lock firstReqDone so only the first one will mark firstReqDone
+				mu.Lock()
+				if !firstReqDone {
+					firstReqDone = true
+					mu.Unlock()
+					go func(r *spaceRequest) {
+						c.makeRequest(r)
+						close(c.run.created) // close this channel to signal that other requests can proceed
+					}(req)
+
+				} else {
+					mu.Unlock()
+
+					// If this is not the first request, wait for the run to be created
+					<-c.run.created
+					go c.makeRequest(req)
+				}
+			}
+		}()
+	}
+}
+
+func (c *spacesClient) makeRequest(req *spaceRequest) {
+	if !c.enabled {
+		return
+	}
+
+	// The runID is required for POST task requests and PATCH run request
+	// so we have to construct it lazily for those requests.
+	// We construc this first in makeRequest, because if makeURL fails, it's likely
+	// because we don't have a runID, which means that the first POST run failed. By checking this
+	// up front, we can avoid duplicate error messages for things like missing spaceID / linking.
+	//
+	// TODO: the purpose of this check up front is just to make sure runID is available for the
+	// requests that need it. Maybe we can leverage the c.run.created channel or another channel for
+	// this so it's more explicit?
+	if req.makeURL != nil {
+		if err := req.makeURL(req, c.run); err != nil {
+			c.errors = append(c.errors, err)
+			return
+		}
+	}
+
+	if c.spaceID == "" {
+		c.errors = append(c.errors, req.error("No spaceID found"))
+		return
+	}
+
+	if !c.api.IsLinked() {
+		c.errors = append(c.errors, req.error("Repo is not linked to a Space. Run `turbo link --target=spaces` first"))
```

Comment:
Are we going to end up with a flood of these errors?

Reply:
No, I have a janky check that prevents the flood of errors, but I think it's a bit brittle, since it relies on an internal detail of how the request url is calculated for requests after the first one.